### PR TITLE
xds: fix ipFamily always nil

### LINKED ISSUE #4761: xdsClusterArgs.ipFamily always nil
https://github.com/envoyproxy/gateway/blob/a924cec6cc3a1d8cb78be06b61b6d238472c199b/internal/xds/translator/cluster.go#L60

### PR DESCRIPTION
**Pull request by juwon8891**
chore: fix ipFamily always nil

**What this PR does / why we need it**:
This PR improves IP family configuration handling across different listener types and updates the DNS lookup strategy for better dual-stack support.
- Add consistent IP family handling across all listener types

**Which issue(s) this PR fixes**:
Fixes #4761

### CODE REVIEW

File: internal/xds/translator/translator.go [hunk 1]
```diff
@@ -463,6 +463,7 @@ func (t *Translator) addRouteToRouteConfig(
 			ea := &ExtraArgs{
 				metrics:       metrics,
 				http1Settings: httpListener.HTTP1,
+				ipFamily:      httpListener.IPFamily,
```
Comment (arkodg) [inline internal/xds/translator/translator.go:466]: this is incorrect, you are using the listener level `IPFamily` value in the cluster
you should be deriving this value using a new `IPFamily` field in 
https://github.com/envoyproxy/gateway/blame/48a0310082ceeaf4d3af17e6b6844c7d0fb5db0b/internal/ir/xds.go#L1297 instead and this should be set in the gateway api layer using the Backend Service / EndpointSlice info

Comment (juwon8891) [reply internal/xds/translator/translator.go]: Thanks for the feedback! I’ve updated the code to use a new IPFamily field

File: internal/gatewayapi/route.go [hunk 2]
```diff
@@ -1261,9 +1261,11 @@ func (t *Translator) processDestination(backendRefContext BackendRefContext,
 			Protocol:    protocol,
 			Endpoints:   endpoints,
 			AddressType: addrType,
+			IPFamily:    getIPFamily(envoyProxy),
```
Comment (arkodg) [inline internal/gatewayapi/route.go:1264]: `getIPFamily` gets the value from the `envoyProxy` which is the listener's IP Family, not the backend's. This value should be coming from https://kubernetes.io/docs/concepts/services-networking/dual-stack/#dual-stack-service-configuration-scenarios

Comment (juwon8891) [reply internal/gatewayapi/route.go]: Thanks for the feedback! I’ve updated the code to use a new IPFamily field
cc @zirain

File: internal/gatewayapi/helpers.go [hunk 3]
```diff
@@ -610,20 +610,56 @@ func setIfNil[T any](target **T, value *T) {
...
+		case corev1.IPv6Protocol:
+			return ptr.To(ir.IPv6)
+		}
+	}
+
+	return nil
+}
+
+// getEnvoyIPFamily returns the IPFamily configuration from EnvoyProxy
+func getEnvoyIPFamily(envoyProxy *egv1a1.EnvoyProxy) *ir.IPFamily {
```
Comment (zirain) [inline internal/gatewayapi/helpers.go:650]: QQ: why we need another enum? cannot we reuse egv1a1.IPFamily?

File: internal/xds/translator/translator.go [hunk 4]
```diff
@@ -465,6 +465,10 @@ func (t *Translator) addRouteToRouteConfig(
 				http1Settings: httpListener.HTTP1,
 			}
 
+			if len(httpRoute.Destination.Settings) > 0 {
```
Comment (zirain) [inline internal/xds/translator/translator.go:468]: is it possible to have different IPFamily in destination settings?

File: internal/gatewayapi/route.go [hunk 5]
```diff
@@ -1313,6 +1314,7 @@ func (t *Translator) processDestination(backendRefContext BackendRefContext,
 	}
 
 	ds.Weight = &weight
+	ds.IPFamily = getServiceIPFamily(resources.GetService(backendNamespace, string(backendRef.Name)))
```
Comment (zirain) [inline internal/gatewayapi/route.go:1317]: better to move to L1282, only service support IPFamily now.

File: internal/gatewayapi/helpers.go (hunk 3 above)
Comment (juwon8891) [reply internal/gatewayapi/helpers.go]: Thanks for the feedback! I’ve updated the code.

File: internal/gatewayapi/route.go (hunk 5 above)
Comment (juwon8891) [reply internal/gatewayapi/route.go]: Thanks for the feedback! I’ve updated the code.

File: internal/xds/translator/translator.go (hunk 4 above)
Comment (juwon8891) [reply internal/xds/translator/translator.go]: Yes, it's possible to have different IPFamily settings in destination configurations. I've also added test cases to confirm this behavior.

Comment (zirain) [reply internal/xds/translator/translator.go]: should can complete `IPFamily` from all `destinations`:
1. if all v4, use v4
2. if all v6, use v6
3. if has both, use Dual?

cc @arkodg , WDYT?

Comment (arkodg) [reply internal/xds/translator/translator.go]: sure SGTM @zirain

Comment (zirain) [reply internal/xds/translator/translator.go]: @juwon8891 can you rebase and fix this?

Comment (juwon8891) [reply internal/xds/translator/translator.go]: Sure I rebased.

File: internal/gatewayapi/helpers.go [hunk 6]
```diff
@@ -610,20 +610,56 @@ func setIfNil[T any](target **T, value *T) {
...
+// getServiceIPFamily returns the IP family configuration from a Kubernetes Service
+// following the dual-stack service configuration scenarios:
+// https://kubernetes.io/docs/concepts/services-networking/dual-stack/#dual-stack-service-configuration-scenarios
+//
+// The IP family is determined in the following order:
+// 1. Service.Spec.IPFamilyPolicy == RequireDualStack -> DualStack
+// 2. Service.Spec.IPFamilies length > 1 -> DualStack
+// 3. Service.Spec.IPFamilies[0] -> IPv4 or IPv6
+// 4. nil if not specified
+func getServiceIPFamily(service *corev1.Service) *egv1a1.IPFamily {
```
Comment (zirain) [inline internal/gatewayapi/helpers.go:622]: it would be better if you can an ut for this.

File: internal/xds/translator/translator.go [hunk 7]
```diff
@@ -464,6 +464,11 @@ func (t *Translator) addRouteToRouteConfig(
 			ea := &ExtraArgs{
 				metrics:       metrics,
 				http1Settings: httpListener.HTTP1,
+				ipFamily:      httpRoute.Destination.Settings[0].IPFamily,
+			}
+
+			if len(httpRoute.Destination.Settings) > 1 {
+				ea.ipFamily = determineIPFamily(httpRoute.Destination.Settings)
```
Comment (zirain) [inline internal/xds/translator/translator.go:471]: merge this into `determineIPFamily`

File: internal/gatewayapi/helpers.go (hunk 6 above)
Comment (juwon8891) [reply internal/gatewayapi/helpers.go:622]: Thank you for your feedback! I've incorporated the suggested changes.

File: internal/xds/translator/translator.go (hunk 7 above)
Comment (juwon8891) [reply internal/xds/translator/translator.go]: Thank you for your feedback! I've incorporated the suggested changes.

File: internal/gatewayapi/listener.go [hunk 8]
```diff
@@ -118,17 +118,14 @@ func (t *Translator) ProcessListeners(gateways []*GatewayContext, xdsIR resource
 						Address:  address,
 						Port:     uint32(containerPort),
 						Metadata: buildListenerMetadata(listener, gateway),
-						IPFamily: getIPFamily(gateway.envoyProxy),
+						IPFamily: getEnvoyIPFamily(gateway.envoyProxy),
```
Comment (zirain) [inline internal/gatewayapi/listener.go:121]:
```suggestion
						IPFamily: ipFamily,
```

File: internal/gatewayapi/listener.go [hunk 9]
```diff
@@ -144,7 +141,7 @@ func (t *Translator) ProcessListeners(gateways []*GatewayContext, xdsIR resource
 						Name:     irListenerName(listener),
 						Address:  address,
 						Port:     uint32(containerPort),
-						IPFamily: getIPFamily(gateway.envoyProxy),
+						IPFamily: getEnvoyIPFamily(gateway.envoyProxy),
```
Comment (zirain) [inline internal/gatewayapi/listener.go:144]:
```suggestion
						IPFamily: ipFamily,
```

**Review by zirain — COMMENTED** [review]
small nit, otherwise LGTM.

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zirain — APPROVED** [review]

**Pull request merged**